### PR TITLE
Feature/virtualized scrolling

### PR DESCRIPTION
**Pull request by ChawinTan**
## What's new

Related to #492

Virtualization of door, lifts and workcell panels to only render what is displayed when scrolling through a list of items. This will improve loading times when there are a lot of items to load on huge maps.

Introduce a new package `react-window`.

### Test results

<b>Without virtualization</b>

| Scripting (ms) | Rendering (ms) | Painting (ms) |
| --- | --- | --- |
| 18372 | 828 | 143 |
| 18400 | 815 | 134 |
| 19057 | 780 | 134 |

<b>Average</b>
1) Scripting - 18.6 seconds
2) Rendering - 0.8 seconds
3) Painting - 0.13 seconds

<b>With Virtualization</b>

| Scripting (ms) | Rendering (ms) | Painting (ms) |
| --- | --- | --- |
| 16009 | 322 | 52 |
| 14872 | 346 | 69 |
| 15693 | 351 | 58 |

<b>Average</b>
1) Scripting - 15.5 seconds
2) Rendering - 0.34 seconds
3) Painting - 0.06 seconds

The tests are done with 6x cpu throttling, from the time i refresh the page to when it finishes loading.

On average, virtualization improves scripting by 3 seconds and more than halve the rendering and painting times.

## Self-checks

- [ ] I have prototyped this new feature (if necessary) on Figma 
- [x] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behavior of the components that interact with the backend, with an e2e test

## Discussion

The downside of using such a virtualization package is that the dimensions of the container and its items has to be fixed. In particular, the width of the table cells in table components are hard coded base on what I inspect on the main branch, else the table wouldn't align properly.

### CODE REVIEW

File: packages/react-components/lib/doors/door-panel.tsx [hunk 1]
```diff
@@ -14,16 +14,26 @@ import React from 'react';
...
+}
+
+export interface DoorGridRendererProps extends GridChildComponentProps {
+  data: DoorGridDataProps;
+}
+
+export interface DoorcellProps {
   door: DoorData;
-  doorMode?: number;
+  doorState?: RmfModels.DoorState;
```
Comment (koonpeng) [inline packages/react-components/lib/doors/door-panel.tsx:36]: Using an object here breaks memo as each update is a different object, revert back to `doorMode` or add custom `React.memo` logic.

File: packages/react-components/lib/doors/door-table.tsx [hunk 2]
```diff
@@ -10,20 +10,48 @@ import {
...
+import clsx from 'clsx';
+import AutoSizer from 'react-virtualized-auto-sizer';
 
 export interface DoorTableProps {
   doors: DoorData[];
   doorStates: Record<string, RmfModels.DoorState>;
   onDoorControlClick?(event: React.MouseEvent, door: RmfModels.Door, mode: number): void;
 }
 
+export interface DoorFixListDataProps extends DoorTableProps {
```
Comment (koonpeng) [inline packages/react-components/lib/doors/door-table.tsx:24]:
```suggestion
interface DoorFixedListData extends DoorTableProps {
```

No export because this shouldn't be used by other components.
No props suffix because this is not used as a prop directly.

File: packages/react-components/lib/lifts/lift-panel.tsx [hunk 3]
```diff
@@ -35,12 +37,17 @@ export interface LiftPanelProps {
...
+  data: LiftGridDataProps;
+}
+
 export interface LiftCellProps {
   lift: RmfModels.Lift;
-  doorState?: number;
-  motionState?: number;
-  destinationFloor?: string;
-  currentFloor?: string;
+  liftState?: RmfModels.LiftState;
```
Comment (koonpeng) [inline packages/react-components/lib/lifts/lift-panel.tsx:50]: Same comment as door.

File: packages/react-components/lib/lifts/lift-table.tsx [hunk 4]
```diff
@@ -36,12 +53,19 @@ export interface LiftTableProps {
...
+  data: LiftFixListDataProps;
+}
+
 export interface LiftRowProps {
   lift: RmfModels.Lift;
-  doorState?: number;
-  destinationFloor?: string;
-  currentFloor?: string;
-  currentMode?: number;
+  liftState: RmfModels.LiftState;
```
Comment (koonpeng) [inline packages/react-components/lib/lifts/lift-table.tsx:66]: Same comment as door.

File: packages/react-components/lib/workcells/workcell-panel.tsx [hunk 5]
```diff
@@ -5,17 +5,30 @@ import { Dispenser } from 'api-client';
...
+
+export interface WorkcellGridRendererProps extends GridChildComponentProps {
+  data: WorkcellGridDataProps;
+}
+
 export interface WorkcellCellProps {
   workcell: Dispenser;
-  requestGuidQueue?: string[];
-  secondsRemaining?: number;
+  workCellState?: RmfModels.DispenserState;
```
Comment (koonpeng) [inline packages/react-components/lib/workcells/workcell-panel.tsx:31]: Same comment as door.

File: packages/react-components/lib/workcells/workcell-table.tsx [hunk 6]
```diff
@@ -14,93 +17,305 @@ const useStyles = makeStyles((theme) => ({
...
+  data: WorkcellFixListDataProps;
+}
+
 export interface WorkcellRowProps {
+  width: number;
   workcell: Dispenser;
-  mode?: number;
-  requestGuidQueue?: string[];
-  secondsRemaining?: number;
+  workcellState: RmfModels.DispenserState;
```
Comment (koonpeng) [inline packages/react-components/lib/workcells/workcell-table.tsx:52]: Same comment as door.

**Review by koonpeng — CHANGES_REQUESTED** [review]

File: packages/react-components/lib/doors/door-panel.tsx [hunk 7]
```diff
@@ -14,16 +14,26 @@ import React from 'react';
...
   doorStates: Record<string, RmfModels.DoorState>;
   onDoorControlClick?(event: React.MouseEvent, door: RmfModels.Door, mode: number): void;
 }
 
-export interface DoorInfoProps {
+export interface DoorGridDataProps extends DoorPanelProps {
+  columnCount: number;
+}
+
+export interface DoorGridRendererProps extends GridChildComponentProps {
```
Comment (koonpeng) [inline packages/react-components/lib/doors/door-panel.tsx:30]: No export because this is not meant to be used by other components.

File: packages/react-components/lib/doors/door-table.tsx [hunk 8]
```diff
@@ -108,28 +192,101 @@ export const DoorTable = ({
...
+        return (
+          <Table stickyHeader size="small" aria-label="door-table" component="div">
+            <TableHead component="div">
+              <TableRow component="div" className={classes.tableRow}>
+                <TableCell
+                  component="div"
+                  variant="head"
+                  className={classes.tableCell}
+                  style={{
+                    minWidth: width * doorTableCellConfig.doorName - 40,
```
[the remark this answers is not in the document]
Comment (koonpeng) [reply packages/react-components/lib/doors/door-table.tsx:240]: Would this work? https://codesandbox.io/s/wizardly-snow-4391f?file=/index.js

The `List` would set the width, and each row just uses flex values to control the width of each column.

[the remark this answers is not in the document]
Comment (ChawinTan) [reply packages/react-components/lib/doors/door-table.tsx]: It works! 

I think my initial gripe was that the table headers and table cells were not perfectly align when the scrollbar showed up. The alignment is just slightly off without the calculations, but I think it is ok.

The `Autosizer` is still needed to provide a width dynamically to the `List` itself.

File: packages/react-components/lib/utils/item-table.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+export function tableCellStyle(flexGrow: string) {
```
Comment (koonpeng) [inline packages/react-components/lib/utils/item-table.ts:1]: Suggest to make this a css class, the properties can be overwritten by making sure the order of the classes in the component is correct.

Also, to not confuse this with other tables, I suggest to name this `useFixedTableCellStyles` or something similar.

**Review by koonpeng — COMMENTED** [review]
Feel free to merge after `tableCellStyle` is addessed.
